### PR TITLE
Add api-auth gem to handle auth between apps

### PR DESCRIPTION
**Pull request by MemunaHaruna**
This adds the api-auth gem and code for authentication and communication between internal applications.

### CODE REVIEW

File: db/migrate/20200326213739_create_api_access_credentials.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class CreateApiAccessCredentials < ActiveRecord::Migration[6.0]
+  def change
+    create_table :api_access_credentials do |t|
```
Comment (dcassanego) [inline db/migrate/20200326213739_create_api_access_credentials.rb]: one thing that i think you should do here (and I have a note to do for the other apps as well, as I should have done it there).  put a column on this table that is like `client_name` or something like that.  just a string so you know who this credential is for.

Comment (MemunaHaruna) [reply db/migrate/20200326213739_create_api_access_credentials.rb]: @dcassanego does the `client_name` represent app identifiers e.g 'galapagos', 'datalake_api' etc? If yes, will the client name be passed (in this case, to Scivias) during the request for confirmation?

I also wanted to confirm the role of the api_access_credential table in the auth process.
Here, I have populated the table with a default access_id and secret_key which I then stored in Galapagos credentials file to be passed to Scivias as request params. Is that the right flow?

File: db/migrate/20200326213739_create_api_access_credentials.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      t.string :access_id
+      t.string :secret_key
+      t.string :client_name
+
+      t.index :access_id
+      t.timestamps
+    end
+
+    ApiAccessCredential.reset_column_information
+    ApiAccessCredential.create!(access_id: rand(10_000...9_999_999),
```
Comment (dcassanego) [inline db/migrate/20200326213739_create_api_access_credentials.rb]: Isolate seed data

File: app/controllers/api_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    render_exception_with_code(exception, :not_found)
+  end
+
+  rescue_from ActionController::ParameterMissing, ActionController::BadRequest do |exception|
+    render_exception_with_code(exception, :bad_request)
+  end
+
+  private #####################################################################################
+
+  def print_auth_debugging(request)
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Move model logic into model (request use_count > 4)

Comment (dcassanego) [inline app/controllers/api_controller.rb]: Assignment Branch Condition size for print_auth_debugging is too high. [<6, 37, 0> 37.48/15]

Comment (dcassanego) [inline app/controllers/api_controller.rb]: Method has too many lines. [21/15]

File: app/controllers/api_controller.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
+class ApiController < ApplicationController
+  before_action :api_authenticate
+
+  def api_authenticate
+    # Uncomment when debugging auth issues
+    # print_auth_debugging(request)
+    @current_account = ApiAccessCredential.find_by_access_id(ApiAuth.access_id(request))
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Use `find_by` instead of dynamic `find_by_access_id`.

File: app/controllers/api_controller.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  rescue_from ActionController::ParameterMissing, ActionController::BadRequest do |exception|
+    render_exception_with_code(exception, :bad_request)
+  end
+
+  private #####################################################################################
+
+  def print_auth_debugging(request)
+    access_id       = ApiAuth.access_id(request)
+    current_account = ApiAccessCredential.find_by_access_id(access_id)
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Use `find_by` instead of dynamic `find_by_access_id`.

File: app/controllers/api_controller.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    auth_header     = ApiAuth::Headers.new(request)
+    ad_request      = ApiAuth::RequestDrivers::ActionDispatchRequest.new(request)
+
+    digest  = OpenSSL::Digest.new('sha1')
+    sig     = Base64.strict_encode64(OpenSSL::HMAC.digest(digest, current_account.secret_key, auth_header.canonical_string))
+
+    puts 'HEADERS ###################################'
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    request.headers.each {|k, v| puts "#{k} :: #{v}" }
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Space between { and | missing.

Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    puts '###########################################'
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    puts "Request Type :: #{request.class.to_s}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

Comment (dcassanego) [inline app/controllers/api_controller.rb]: Redundant use of `Object#to_s` in interpolation.

File: app/controllers/api_controller.rb [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    puts "Client ID :: #{access_id}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    puts '-------------'
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    puts "RAW BODY :: #{request.raw_post}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    puts "RAW BODY NIL? :: #{request.raw_post.nil?}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    puts "RAW BODY BLANK? :: #{request.raw_post.blank?}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    puts "Content MD5 :: #{ad_request.content_md5}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    puts "Computed MD5 :: #{ad_request.calculated_md5}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    puts "MD5 Mismatch :: #{auth_header.md5_mismatch?}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+    puts '-------------'
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    puts "Canonical String :: #{auth_header.canonical_string}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: app/controllers/api_controller.rb [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+    puts "Computed Signature :: #{sig}"
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Do not write to stdout. Use Rails's logger if you want to log.

File: test/models/api_access_credential_test.rb [hunk 21]
```diff
@@ -0,0 +1,11 @@
+require 'test_helper'
+
+class ApiAccessCredentialTest < ActiveSupport::TestCase
+  test "valid record" do
```
Comment (dcassanego) [inline test/models/api_access_credential_test.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: test/models/api_access_credential_test.rb [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+    new_credential = ApiAccessCredential.new secret_key: '1234'
+    assert_not new_credential.valid?
+
+    new_credential = ApiAccessCredential.new secret_key: 'abcd', access_id: '123455',
+                      client_name: 'galapagos'
```
Comment (dcassanego) [inline test/models/api_access_credential_test.rb]: Align the keys of a hash literal if they span more than one line.

File: app/controllers/api_controller.rb [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    render_exception_with_code(exception, :not_found)
+  end
+
+  rescue_from ActionController::ParameterMissing, ActionController::BadRequest do |exception|
+    render_exception_with_code(exception, :bad_request)
+  end
+
+  private #####################################################################################
+
+  def print_auth_debugging(request)
```
Comment (dcassanego) [inline app/controllers/api_controller.rb]: Move model logic into model (rails use_count > 4)

Comment (dcassanego) [inline app/controllers/api_controller.rb]: Assignment Branch Condition size for print_auth_debugging is too high. [<6, 51, 0> 51.35/15]

**Pull request merged**